### PR TITLE
Introduce ScanAll operation

### PR DESCRIPTION
**Pull request by brfrn169**
This PR introduces a `ScanAll` operation that scans all the data of a database. This PR doesn't contain the actual implementation of `ScanAll` but it contains only  the `ScanAll` operation class.

I added the `ScanAll` as a subclass of `Scan` because I think that minimizes the total change cost.

I assume the `ScanAll` operation is used with the existing `scan()` API as follows:
```java
Scanner scanner =
    storage.scan(
        new ScanAll().withProjection("col1").withLimit(10).forNamespace("ns").forTable("tbl"));
```
Currently `ScanAll` only supports specifying projections and limit.

Please take a look!

### CODE REVIEW

**Review by feeblefakie — APPROVED** [review]
LGTM! Thank you!

File: core/src/main/java/com/scalar/db/io/Key.java [hunk 1]
```diff
@@ -586,6 +586,15 @@ public static Key ofBlob(String columnName, ByteBuffer value) {
     return new Key(columnName, value);
   }
 
+  /**
+   * Create an empty {@code Key} object
+   *
+   * @return an empty {@code Key} object
+   */
+  public static Key of() {
```
Comment (Torch3333) [inline core/src/main/java/com/scalar/db/io/Key.java:594]: Would `Key.empty()` be appropriate ?

File: core/src/main/java/com/scalar/db/api/ScanAll.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+  @Override
+  public Scan withEnd(Key clusteringKey) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public Scan withEnd(Key clusteringKey, boolean inclusive) {
+    throw new UnsupportedOperationException();
+  }
```
Comment (Torch3333) [inline core/src/main/java/com/scalar/db/api/ScanAll.java:45]: If I understood correctly, you would like to implement these methods for ScanAll later on. 
Though intuitively, as a user, If I want only to retrieve a subset of all the database entries, wouldn't I want to filter on the entire primary key and not only on the clustering key only?

**Review by Torch3333 — COMMENTED** [review]
Thank you. I left some comments.

File: core/src/main/java/com/scalar/db/api/ScanAll.java (hunk 2 above)
Comment (brfrn169) [reply core/src/main/java/com/scalar/db/api/ScanAll.java:45]: In this `ScanAll` operation, we won't support a clustering key range scan. Instead, we can add other methods filtering on arbitrary columns in the future. At this point, we support only specifying projections and limit.

File: core/src/main/java/com/scalar/db/io/Key.java (hunk 1 above)
Comment (brfrn169) [reply core/src/main/java/com/scalar/db/io/Key.java:594]: In `ImmutableXXX` in Guava, we can create empty collections with `of()` without arguments. I followed this naming. What do you think?

**Review by brfrn169 — COMMENTED** [review]
@Torch3333 Thank you for taking a look at this PR. I reply for your comments. Please take a look when you have time!

File: core/src/main/java/com/scalar/db/api/ScanAll.java (hunk 2 above)
Comment (Torch3333) [reply core/src/main/java/com/scalar/db/api/ScanAll.java:45]: All right. 
I wouldn't say I like that we have to expose the non-implemented `withEnd` and `withStart` methods to the end-user, but if it's a necessary evil to reduce the development cost, then I understand.

File: core/src/main/java/com/scalar/db/io/Key.java (hunk 1 above)
Comment (Torch3333) [reply core/src/main/java/com/scalar/db/io/Key.java:594]: All right. If it's a well known usage, that's good.

**Review by Torch3333 — APPROVED** [review]
Thank you!
LGTM

**Pull request merged**